src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java

```diff
@@ -56,57 +71,186 @@
+    {
+        this.scheduler = scheduler;
+    }
+
+    public void setTestMode(boolean testMode)
+    {
+        this.testMode = testMode;
+    }
+
     /**
      *
-     * Register data collector with this service.
-     * The registered collectors will be called to provide heartbeat data.
+     * Register data collector with this service and start the schedule.
+     * The registered collector will be called to provide heartbeat data.
+     * The collector must be unique with using the collector id.
      *
      * @param collector collector to register
      */
     @Override
     public void registerCollector(HBBaseDataCollector collector)
     {
+        for(HBBaseDataCollector col : collectors)
+        {
+            if(col.getCollectorId().equals(collector.getCollectorId()))
+            {
+                if (logger.isDebugEnabled())
+                {
+                    logger.debug("Didn't register collector because it already exists: " + collector.getCollectorId());
+                }
+                return;
+            }
+        }
         this.collectors.add(collector);
+        try
+        {
+            scheduleCollector(collector);
+        }
+        catch (Exception e)
+        {
+            logger.error("Unable to schedule heart beat", e);
+        }
     }
 
     /**
-     *  Collects and sends data for all registered collectors using the provided sender service.
+     * Listens for license changes.  If a license is change or removed, the heartbeat job is rescheduled.
      */
-    @Override
-    public void collectAndSendData()
+    public synchronized void onLicenseChange(LicenseDescriptor licenseDescriptor)
+    {
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("Update license called");
+        }
+
+        boolean newEnabled = !licenseDescriptor.isHeartBeatDisabled();
+
+        if (newEnabled != this.enabled)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("State change of heartbeat");
+            }
+            enable(newEnabled);
+            restartAllCollectorSchedules();
+        }
+    }
+
+    /**
+     * License load failure resets the heartbeat back to the default state
+     */
+    public synchronized void onLicenseFail()
+    {
+        boolean newEnabled = isEnabledByDefault();
+
+        if (newEnabled != this.enabled)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("State change of heartbeat");
+            }
+            enable(newEnabled);
+            restartAllCollectorSchedules();
+        }
+    }
+
+    private void restartAllCollectorSchedules()
+    {
+        try
+        {
+            for(HBBaseDataCollector collector : collectors)
+            {
+                scheduleCollector(collector);
+            }
+        }
+        catch (Exception e)
+        {
+            logger.error("Unable to schedule heart beat", e);
+        }
```

Comment:
Move the try catch inside of the for loop and log exactly which collector failed to schedule/unschedule its job. 
As it is now we will break out of the for loop at the first iteration that throws an exception.
The behaviour we probably want is to log when a collector fails to schedule/unschedule job and to continue the for loop for the other collectors.
Would we want some kind of transaction for this?